src/cpp/session/modules/environment/SessionEnvironment.cpp

```diff
@@ -119,95 +120,174 @@ bool handleRBrowseEnv(const core::FilePath& filePath)
+         return true;
+
+      if (hasExternalPointer(EXTPTR_TAG(obj), nullPtr, visited))
+         return true;
    }
-   else
+
+   // check attributes, this includes slots for S4 objects
+   if (attributesHasExternalPointer(obj, nullPtr, visited))
+      return true;
+
+   switch(TYPEOF(obj))
    {
-      // not S4, coerce to environment
-      SEXP envir = R_NilValue;
-      if (TYPEOF(obj) == ENVSXP)
+      case ENVSXP: 
       {
-         // we were given a primitive environment (ENVSXP)
-         envir = obj;
+         if (envHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
       }
-      else
+      case VECSXP:
+      case EXPRSXP:
       {
-         // convert the passed environment into a primitive environment; this is required so that
-         // e.g. reference objects that subclass 'environment' can be introspected below
-         Error error = r::sexp::asPrimitiveEnvironment(obj, &envir, &rProtect);
-         if (error)
-         {
-            // can't search in here
-            return false;
-         }
+         if (listHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
       }
-
-      r::sexp::listEnvironment(envir,
-                               true,  // include all values
-                               false, // don't include last dot
-                               &rProtect, &vars);
-   }
-
-   // check for external pointers
-   for (std::vector<r::sexp::Variable>::iterator it = vars.begin(); it != vars.end(); it++)
-   {
-      if (r::sexp::isExternalPointer(it->second) &&
-          r::sexp::isNullExternalPointer(it->second) == nullPtr)
+         
+      case LISTSXP:
+      case LANGSXP:
       {
-         return true;
+         if (pairlistHasExternalPointer(obj, nullPtr, visited))
+            return true;
+         break;
       }
+         
+      case WEAKREFSXP:
+      {
+         SEXP key = VECTOR_ELT(obj, 0); // VECTOR_ELT(., 0) == WEAKREF_KEY (memory.c)
+         if (key != R_NilValue)
+         {
+            if (hasExternalPointer(key, nullPtr, visited))
+               return true;
+
+            // only consider the value if the key is not NULL
+            // VECTOR_ELT(., 0) == WEAKREF_VALUE (memory.c)
+            if (hasExternalPointer(VECTOR_ELT(obj, 1), nullPtr, visited))
+               return true;
 
-      if (r::sexp::isPrimitiveEnvironment(it->second) || TYPEOF(it->second) == S4SXP)
+            // WEAKREF_FINALIZER ?
+         }
+         break;
+      }
+      case PROMSXP: 
       {
-         // if this object is itself an environment, check it recursively for external pointers.
-         // (we do this only if there's sufficient recursion depth remaining)
-         if (hasExternalPtrImpl(it->second, nullPtr, visited))
+         SEXP value = PRVALUE(var);
+         if (value != R_UnboundValue && hasExternalPointer(value, nullPtr, visited))
             return true;
+         
```

Comment:
Should this look for external pointers in `PRCODE()` and/or `PRENV()` when the promise wasn't yet evaluated ?